### ISSUE TITLE
GridMap only supports 16bit IDs while MeshLibrary supports 32bit IDs

### ISSUE THREAD

**Issue by wd40bomber7** [godotengine/godot#43836]
**Godot version:** v3.2.3.stable.mono.official

**OS/device including version:** Microsoft Windows 19041.630

**Issue description:** I am working on a very simple game and decided I wanted to use gridmap to paint the terrain. I wrote a tiny bit of C# script to take a list of textures and build simple 3d mesh tiles using them to create my mesh library. For the MeshLibrary IDs I decided to just hash the texture names so the IDs would stay consistent every time I regenerated the MeshLibrary. This is the important bit because I generated those ids like this:
`Math.Abs(itemName.GetHashCode())`

This generated entirely valid MeshLibrary entries:
![image](https://user-images.githubusercontent.com/2555227/100156739-4be47900-2e5e-11eb-8f81-2a63ca14b411.png)

However when I went to actually draw my grid tiles nothing happened (looking like this):
![GridMapBroken](https://user-images.githubusercontent.com/2555227/100156839-7a625400-2e5e-11eb-992d-218359b92dfb.gif)

The problem is this code from [gridmap.cpp](https://github.com/godotengine/godot/blob/90edd839d5f67d4252923a720feeef6ad0d76107/modules/gridmap/grid_map.cpp):

`
void GridMap::set_cell_item(const Vector3i &p_position, int p_item, int p_rot) {
   // ... Lots of unrelated stuff ...
	Cell c;
	c.item = p_item;
	c.rot = p_rot;

	cell_map[key] = c;
}
`

Cell is defined like this:
`
	union Cell {
		struct {
			unsigned int item : 16;
			unsigned int rot : 5;
			unsigned int layer : 8;
		};
		uint32_t cell;
                // Constructor omitted...
	};
`
And there's the problem. Item is bitmapped to only the first 16 bits. This means that the latter 16 bits are just chopped right off without any error or warning! The shorter and completely wrong ID is then stored in the gridmap.

In summary:

- MeshLibrary supports 32 bit positive IDs
- GridMap only supports 16 bit IDs
- GridMap takes 32 bit IDs and then cuts off the top 16 bits without any warning or error

**Minimal reproduction project: [GridMapKeyLimitation.zip](https://github.com/godotengine/godot/files/5592986/GridMapKeyLimitation.zip)**

**Comment by ATHellboy** [comment]
Any plans on supporting 32bit IDs for GridMap ?
